californium-core/src/main/java/org/eclipse/californium/core/network/stack/TcpBlockwiseLayer.java

```diff
@@ -0,0 +1,523 @@
+			final Block2BlockwiseStatus status) {
+
+		BlockOption block2 = request.getOptions().getBlock2();
+		if ((block2 != null) && (block2.getSzx() == BERT_SZX)) {
+
+			synchronized (status) {
+
+				ByteBuffer responseBlockBuilder = ByteBuffer.allocate(bertStepSize * BERT_INT_BLOCK_SIZE);
+				int blockNum = block2.getNum();
+				int currentNum = blockNum;
+				boolean hasNextBlock = true;
+				Response block = null;
+
+				for (int i = 0; ((i < bertStepSize) && (hasNextBlock)); i++) {
+					block = status.getNextResponseBlock(block2);
+					hasNextBlock = block.getOptions().getBlock2().isM();
+					currentNum = currentNum + 1;
+					block2 = new BlockOption(BERT_SZX, hasNextBlock, currentNum);
+					responseBlockBuilder.put(block.getPayload());
+				}
+
+				if (status.isComplete()) {
+					// clean up blockwise status
+					LOGGER.log(Level.FINE, "peer has requested last block of blockwise transfer: {0}", status);
+					clearBlock2Status(key);
+				} else {
+					LOGGER.log(Level.FINE, "peer has requested intermediary block of blockwise transfer: {0}", status);
+				}
+
+				block.setPayload(getBody(responseBlockBuilder));
+				boolean m = block.getOptions().getBlock2().isM();
+				block.getOptions().setBlock2(BERT_SZX, m, blockNum);
+				exchange.setCurrentResponse(block);
+				lower().sendResponse(exchange, block);
+			}
+		} else {
+			super.handleInboundRequestForNextBlock(exchange, request, key, status);
+		}
+	}
+
+	/**
+	 * Send request to peers.Handling is done here only when BERT option is
+	 * enabled otherwise handling is done in the BlockwikseLayer.
+	 */
+	public void sendRequest(final Exchange exchange, final Request request) {
+
+		Request requestToSend = request;
+		if (isBertEnabled) {
+			Request firstRequest = request;
+			if (isTransparentBlockwiseHandlingEnabled()) {
+
+				BlockOption block2 = request.getOptions().getBlock2();
+				if (block2 != null && block2.getNum() > 0) {
+					// This is the case if the user has explicitly added a block
+					// option
+					// for random access.
+					// Note: We do not regard it as random access when the block
+					// number is 0.
+					// This is because the user might just want to do early
+					// block
+					// size negotiation but actually want to retrieve the whole
+					// body by means of
+					// a transparent blockwise transfer.
+					LOGGER.fine("outbound request contains block2 option, creating random-access blockwise status");
+					addRandomAccessBlock2Status(exchange, request);
+					handleRandomBlockAccess(exchange, request, block2.getNum());
+				} else if (requiresBlockwise(request)) {
+					// This must be a large POST or PUT request with BERT
+					// option.
+					ByteBuffer requestBuilder = ByteBuffer.allocate(bertStepSize * BERT_INT_BLOCK_SIZE);
+					boolean hasNextBlock = true;
+					KeyUri key = getKey(exchange, request);
+					firstRequest = startBlockwiseUpload(exchange, request);
```

Comment:
`firstRequest` is only used within this scope, so I would consider to move the line 206 

> Request firstRequest = request;

into this block.